[YU SHUFAN] iP

fans2619 opened the pull request:
### Duke Plus

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke Plus frees your mind of having to remember things you need to do. It's,

- text-based
- fast to learn
- ~~EASY~~ *SUPER* EASY to use

All you need to do is,

1. download it from [here](https://github.com/fans2619/ip/releases/download/v0.1/ip.jar).
2. open a terminal on your computer with JDK installed and navigate to the path where you saved the `ip.jar` file.
3. run `java -jar ip.jar`.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x]  Managing tasks of three types: Todo, Deadline, Event
- [ ] Reminders (coming soon)

------

If you are a Java programmer, you can use it to practice Java too. Refer to the [GitHub page](https://github.com/fans2619/ip) for the source code! Here's the `main` method:

```java
public static void main(String[] args) {
    new Duke("data/duke.txt").run();
}
```

Commit:
Add Gradle support

fans2619 committed:
Implement increments for Level-1: Greet, Echo, Exit

fans2619 committed:
Implement increments for Level-2: Add, List

A new class TaskList is added.
Duke.java is modified.

fans2619 committed:
Implement increments for Level-3: Mark as Done

Two new classes, DukeCore and Task, are added.
DukeCore is the core class of the chatterbot.

fans2619 committed:
Add three new classes: Todo, Deadline, Event

fans2619 committed:
Modify DukeCore and TaskList to work for different types of tasks

fans2619 committed:
Perform an automated text UI testing

fans2619 committed:
Implement increments for Level-5: Handle Errors

fans2619 committed:
Implement increments for Level-6: Delete

fans2619 committed:
Perform an automated text UI testing

fans2619 committed:
Add some test cases

fans2619 committed:
Implement increments for A-Enums: Use Enums

Add an enum class TaskType to represent the three types of task

fans2619 committed:
Implement increments for Level-7: Save

fans2619 committed:
Modify test script

fans2619 committed:
Implement increments for Level-8: Dates and Times

fans2619 committed:
Merge branch 'branch-Level-7'

fans2619 committed:
Merge branch 'branch-Level-8'

fans2619 committed:
DukeDataHandler.java: Fix some bugs related to DateTime

fans2619 committed:
Refactor the whole project to have more OOP

fans2619 committed:
Organise files into packages

fans2619 committed:
Duke.java: Fix wrong package

fans2619 committed:
Add JUnit Tests

fans2619 committed:
Update .gitignore

fans2619 committed:
Add JavaDoc comments

fans2619 committed:
Tweak some code to comply with the required coding standard

fans2619 committed:
Implement increments for Level-9: Find

fans2619 committed:
Merge branch 'branch-A-JavaDoc'

fans2619 committed:
Merge branch 'branch-A-CodingStandard'

fans2619 committed:
Merge branch 'branch-Level-9'

fans2619 committed:
FindCommand.java: Fix some bugs

fans2619 committed:
Parser.java: Make some enhancements

fans2619 committed:
Edit some JavaDoc

fans2619 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

fans2619 committed:
Add Gradle support to the project

fans2619 committed:
Refactor the structure of packages to follow the naming convention

fans2619 committed:
Implement Duke's GUI

Todo: update some JavaDoc

fans2619 committed:
Improve Duke's GUI

Todo: update some JavaDoc

fans2619 committed:
Improve Duke's GUI

Some JavaDoc has been updated at the same time.

fans2619 committed:
Merge branch 'branch-Level-10'

Nature711 reviewed (commented):
LGTM. No violation of coding standard found 👍

tsm1820 commented on src/main/java/fan/cs2103t/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,57 @@
+        try {
+            Task t = taskList.getTasks().get(taskIndex - 1);
+            String description = t.getDescriptionWithStatus();
+            if (taskList.deleteTask(taskIndex - 1)) {
```
Hi, you seem to have followed the coding standards pretty well, so good job on that! As for the deleteTask method here, I am assuming it returns a boolean? But the name of the method does not make it clear that it returns a boolean. Perhaps you could rename it to wasTaskDeleted()?

fans2619 replied:
For this one ah, I will leave it there. I didn't name the method as `wasTaskDeleted()` because returning a boolean value is more like a "side effect" of the method instead of its main purpose. If you check Java Documentation (for example [this](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ArrayList.html#add(E))), you could find a similar case.

tsm1820 commented on src/main/java/fan/cs2103t/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,57 @@
+
+    /**
+     * Executes this command. Changes the completion status of the task with the index to "done".
+     * Displays a message to the user through the specified UI if the status of the task is successfully changed,
```
I may be going too much into the details, but this line seems too long, maybe you can split it and make it a bit shorter?

tsm1820 commented on src/main/java/fan/cs2103t/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,38 @@
+     * @return true if the specified command is an instance of <code>ExitCommand</code>.
+     */
+    public static boolean isExit(Command command) {
+        return command instanceof ExitCommand;
```
Nice use of the instanceof operator here, good job!

tsm1820 commented on src/main/java/fan/cs2103t/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,64 @@
+        String output;
+        StringBuilder result = new StringBuilder();
+        int count = 0;
+        boolean hasFound = false;
```
Again being a bit picky here, but perhaps you could rename hasFound to isFound, cause that is more intuitive considering this is a boolean variable.

fans2619 replied:
Sure will change!

tsm1820 commented on src/main/java/fan/cs2103t/duke/task/TaskList.java:
```diff
@@ -0,0 +1,95 @@
+     * @param task task to be appended to this task list.
+     * @return <code>true</code> if the task is added successfully.
+     */
+    public boolean addTask(Task task) {
```
Again I feel the addTask method name does not clarify that a boolean variable is being returned

tsm1820 reviewed (commented):
LGTM. Overall the code is pretty clean, and the coding standards and code quality are both up to the mark! Good Job! Just some nitty gritty details could be improved upon.

fans2619 committed:
Add two assertions

fans2619 committed:
Merge pull request #1 from fans2619/branch-A-Assertions

Add two assertions

fans2619 committed:
Improve code quality

Many classes contain different messages to be displayed to the users.
Those messages can seem magic and make the code hard to understand.
Let's extract all the messages as constants into a Messages class.
This makes the code clearer and future understanding and modification
easier.

fans2619 committed:
Merge pull request #2 from fans2619/branch-A-CodeQuality

Improve code quality

fans2619 committed:
FindCommand.java: Replace the for loop for searching with stream

fans2619 committed:
Merge pull request #3 from fans2619/branch-A-Streams

FindCommand.java: Replace the for loop for searching with stream

fans2619 committed:
Set up CI

fans2619 committed:
Improve Duke to allow mass Done and Delete operations.

"done" and "delete" commands only allow one number as their argument.

This makes it painful for users when they want to mark several tasks
as done or delete several tasks at one time.

Let's modify DoneCommand and DeleteCommand to allow multiple numbers
as arguments of "done" and "delete" commands. For example,
now users can key in "done 1 2 3 4" or "delete 1 2 3 4".

fans2619 committed:
Merge pull request #4 from fans2619/branch-C-MassOps

Improve Duke to allow mass Done and Delete operations.

fans2619 committed:
Messages.java: Solve a coding standard issue

fans2619 committed:
FindCommand.java: Solve an issue raised by peer PR review

fans2619 committed:
Add a representative screenshot of the product

fans2619 committed:
Remove all CLI outputs

fans2619 committed:
Add a start interface image

fans2619 committed:
FindCommand.java: Make search case-insensitive

fans2619 committed:
Update README.md

fans2619 committed:
Update User Guide

fans2619 committed:
Update User Guide

fans2619 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	docs/README.md

fans2619 committed:
Update User Guide

fans2619 committed:
Update User Guide

fans2619 committed:
Update User Guide

fans2619 committed:
Update User Guide

fans2619 committed:
Update README.md
